From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] wsdapi: IWSDUdpAddress exclusive flag, TTL and alias (3542)

wsdapi: IWSDUdpAddress exclusive flag, TTL and alias

IWSDUdpAddress SetExclusive/GetExclusive (S_OK exclusive, S_FALSE not), SetTTL/GetTTL and SetAlias/GetAlias keep their values.
---
--- a/dlls/wsdapi/address.c
+++ b/dlls/wsdapi/address.c
@@ -35,6 +35,9 @@
     WCHAR            ipv6Address[64];
     WORD             port;
     WSDUdpMessageType messageType;
+    BOOL             exclusive;
+    DWORD            ttl;
+    GUID             alias;
 } IWSDUdpAddressImpl;
 
 static inline IWSDUdpAddressImpl *impl_from_IWSDUdpAddress(IWSDUdpAddress *iface)
@@ -257,14 +260,21 @@
 
 static HRESULT WINAPI IWSDUdpAddressImpl_SetExclusive(IWSDUdpAddress *This, BOOL fExclusive)
 {
-    FIXME("(%p, %d)\n", This, fExclusive);
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p, %d)\n", This, fExclusive);
+
+    impl->exclusive = !!fExclusive;
+    return S_OK;
 }
 
 static HRESULT WINAPI IWSDUdpAddressImpl_GetExclusive(IWSDUdpAddress *This)
 {
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p)\n", This);
+
+    return impl->exclusive ? S_OK : S_FALSE;
 }
 
 static HRESULT WINAPI IWSDUdpAddressImpl_SetMessageType(IWSDUdpAddress *This, WSDUdpMessageType messageType)
@@ -294,26 +304,51 @@
 
 static HRESULT WINAPI IWSDUdpAddressImpl_SetTTL(IWSDUdpAddress *This, DWORD dwTTL)
 {
-    FIXME("(%p, %ld)\n", This, dwTTL);
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p, %ld)\n", This, dwTTL);
+
+    impl->ttl = dwTTL;
+    return S_OK;
 }
 
 static HRESULT WINAPI IWSDUdpAddressImpl_GetTTL(IWSDUdpAddress *This, DWORD *pdwTTL)
 {
-    FIXME("(%p, %p)\n", This, pdwTTL);
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p, %p)\n", This, pdwTTL);
+
+    if (!pdwTTL)
+        return E_POINTER;
+
+    *pdwTTL = impl->ttl;
+    return S_OK;
 }
 
 static HRESULT WINAPI IWSDUdpAddressImpl_SetAlias(IWSDUdpAddress *This, const GUID *pAlias)
 {
-    FIXME("(%p, %s)\n", This, debugstr_guid(pAlias));
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p, %s)\n", This, debugstr_guid(pAlias));
+
+    if (!pAlias)
+        return E_POINTER;
+
+    impl->alias = *pAlias;
+    return S_OK;
 }
 
 static HRESULT WINAPI IWSDUdpAddressImpl_GetAlias(IWSDUdpAddress *This, GUID *pAlias)
 {
-    FIXME("(%p, %p)\n", This, pAlias);
-    return E_NOTIMPL;
+    IWSDUdpAddressImpl *impl = impl_from_IWSDUdpAddress(This);
+
+    TRACE("(%p, %p)\n", This, pAlias);
+
+    if (!pAlias)
+        return E_POINTER;
+
+    *pAlias = impl->alias;
+    return S_OK;
 }
 
 static const IWSDUdpAddressVtbl udpAddressVtbl =
